overlord/snapstate/autorefresh_gating.go

```diff
@@ -33,6 +36,270 @@ import (
 
 var gateAutoRefreshHookName = "gate-auto-refresh"
 
+// cumulative hold time for snaps other than self
+const maxOtherHoldDuration = time.Hour * 48
+
+var timeNow = func() time.Time {
+	return time.Now()
+}
+
+func lastRefreshed(st *state.State, snapName string) (time.Time, error) {
+	var snapst SnapState
+	if err := Get(st, snapName, &snapst); err != nil {
+		return time.Time{}, fmt.Errorf("internal error, cannot get snap %q: %v", snapName, err)
+	}
+	// try to get last refresh time from snapstate, but it may not be present
+	// for snaps installed before the introduction of last-refresh attribute.
+	if snapst.LastRefreshTime != nil {
+		return *snapst.LastRefreshTime, nil
+	}
+	snapInfo, err := snapst.CurrentInfo()
+	if err != nil {
+		return time.Time{}, err
+	}
+	// fall back to the modification time of .snap blob file as it's the best
+	// approximation of last refresh time.
+	fst, err := os.Stat(snapInfo.MountFile())
+	if err != nil {
+		return time.Time{}, err
+	}
+	return fst.ModTime(), nil
+}
+
+type holdState struct {
+	// FirstHeld keeps the time when the given snap was first held for refresh by a gating snap.
+	FirstHeld time.Time `json:"first-held"`
+	// HoldUntil stores the desired end time for holding.
+	HoldUntil time.Time `json:"hold-until"`
+}
+
+func refreshGating(st *state.State) (map[string]map[string]*holdState, error) {
+	// held snaps -> holding snap(s) -> first-held/hold-until time
+	var gating map[string]map[string]*holdState
+	err := st.Get("snaps-hold", &gating)
+	if err != nil && err != state.ErrNoState {
+		return nil, fmt.Errorf("internal error: cannot get snaps-hold: %v", err)
+	}
+	if err == state.ErrNoState {
+		return make(map[string]map[string]*holdState), nil
+	}
+	return gating, nil
+}
+
+// HoldDurationError contains the that error prevents requested hold, along with
+// hold time that's left (if any).
+type HoldDurationError struct {
+	Err          error
+	DurationLeft time.Duration
+}
+
+func (h *HoldDurationError) Error() string {
+	return h.Err.Error()
+}
+
+// HoldError contains the details of snaps that cannot to be held.
+type HoldError struct {
+	SnapsInError map[string]HoldDurationError
+}
+
+func (h *HoldError) Error() string {
+	l := []string{""}
+	for _, e := range h.SnapsInError {
+		l = append(l, e.Error())
+	}
+	return fmt.Sprintf("cannot hold some snaps:%s", strings.Join(l, "\n - "))
+}
+
+func maxAllowedPostponement(gatingSnap, affectedSnap string, maxPostponement time.Duration) time.Duration {
+	if affectedSnap == gatingSnap {
+		return maxPostponement
+	}
+	return maxOtherHoldDuration
+}
+
+// holdDurationLeft computes the maximum duration that's left for holding a refresh
+// given current time, last refresh time, time when snap was first held, maximum
+// duration allowed for the given snap and maximum overall postponement allowed by
+// snapd.
+func holdDurationLeft(now time.Time, lastRefresh, firstHeld time.Time, maxDuration, maxPostponement time.Duration) time.Duration {
```

Comment:
I assume `maxDuration` is the max duration allowed that a snap could be held, correct? So that could be maxOtherHoldDuration or maxPostponementBuffer ?

Reply:
Yes. But it can be maxOtherHoldDuration or maxPostponement or (maxPostponement - maxPostponementBuffer).

Reply:
ah right, thanks for the clarification